### ISSUE TITLE
[1.15.2 pre24] Breaking blocks with a pickaxe causes pattern to fill screen

### ISSUE THREAD

**Issue by AnonymousHacker1279** [sp614x/optifine#4088]
## Description of Issue
When breaking any block with a pickaxe, the block breaking pattern appears on a solid background and fills the screen. Sometimes this occurs shortly after starting to break the block.

![Breaking obsidian with a diamond pickaxe, triggering the effect](https://user-images.githubusercontent.com/48810167/84599068-4b6e9d80-ae3d-11ea-9575-1efdf1049747.png)

## Steps to Reproduce
1. Attempt to break any block with a pickaxe

## OptiFine Version
1.15.2 pre24

## Installation Method
.jar file was placed into the Forge 'mods' folder.

## Fabric/Forge Version
Forge 1.15.2-31.2.19

## Other Installed Mods
[Pastebin](https://pastebin.com/1F4C8vQ4)

## Log Files/Crash Reports
The game doesn't crash, so a crash report wasn't generated.

## F3 Debug Screenshot
![F3 debug screenshot](https://user-images.githubusercontent.com/48810167/84599307-064b6b00-ae3f-11ea-977b-05fdd4fa858a.png)

## Prior Testing
This does not happen when OptiFine is disabled, installed standalone, or installed with forge. Installed texturepacks don't seem to influence this issue (I tried with and without).

Internal shaders were used with all graphics settings set to their max.

It appears to be related to a combination of installed mods but I am unable to determine which ones are problematic. When I tried each other mod by itself and with OptiFine, the issue did not arise.

## Additional Information
For some reason this only seems to happen when the block is being broken by a pickaxe. It didn't happen for me when I mined by hand or with other tools.

**Comment by thepriz** [comment]
This is still happening in 1.15.2 pre26. It was not happening with 1.15.2 pre23 as I was using that before I upgraded.

**Comment by Kynolin** [comment]
I haven't yet fully figured out what does/doesn't cause this, but it happens to both vanilla and mod tools. It seems to be somewhat related to the effectiveness or category of the tool used to break the block.

I'm using FTB Omnia (1.15.2) with Optifine pre24 and the internal shaders. SEUS has the same issue, but it's more of a weird transparent effect (and a bit easier to see through).

Here are some combinations I tried. Yes means the whole screen was covered with the block breaking animation. No means the block breaking animation was on the block itself as normal. Obsidian may be hard to tell with the low mining speed.

wooden pickaxe
- grass: yes
- oak log: yes
- cobblestone: yes
- obsidian: no?

wooden shovel
- grass: yes
- oak log: yes
- cobblestone: yes
- obsidian: no?

wooden axe
- grass: yes
- oak log: yes
- cobblestone: yes
- obsidian: no?

wooden hoe
- grass: no
- oak log: no
- cobblestone: no
- obsidian: no
- farmland: no

Lots of issues with:
- Terra Shatterer (Botania)
- Refined Obsidian Paxel (Mekanism)
- Lapis Lazuli Pickaxe (Mekanism)

No issues so far with:
- iron sword
- blazing immolation blade (Wither Skeleton Tweaks)
- Epic Backpack (Simply Backpacks)
- Wand of the Forest (Botania)

**Comment by Leirday** [comment]
This has been happening to me as well on the pre26 version. In my case, breaking blocks with tools (no punches) makes my screen go black at (almost) the same moment that I start breaking a block and it ends after the block has been broken.

**Comment by sp614x** [comment]
Looks like a mod conflict.
Can you identify the problematic mod by:
- splitting mods in two groups
- testing one group and keeping the group that has problems
- repeat until only one mod except OptiFine is left

**Comment by Kynolin** [comment]
That wasn't fun with 170 mods and dependencies but good call with the binary search approach.

Here's the conflict for me:

preview_OptiFine_1.15.2_HD_U_G1_pre24.jar
cofh_core-1.15.2-0.4.0b.jar

**Comment by AnonymousHacker1279** [comment]
CoFH Core (1.15.2-0.4.0b) also seems to be the problematic mod for me.

**Comment by Kynolin** [comment]
Contributors to CoFH have attributed this issue to Optifine, citing that Optifine is closed source, and this issue was introduced by an Optifine update. I started looking at the source to CoFHCore until realizing it's for 1.12, and their code for 1.15.2 is still private. CoFH is obviously doing something that 169 other mods aren't, and Optifine is doing something in version 24 that has an issue with this, but I'm afraid there's nothing further I can do to help other than hope it gets fixed somewhere.

**Comment by leoshusar** [comment]
I don't know if my problem is the same, but it's happening for me only with shaders.
I am playing FTB Omnia with OptiFine pre28 (happens with pre24 too). When I start digging (grass, crafting table) with hand, everything is ok, but when I use axe, pickaxe, shovel..., my screen goes all black. When I stop digging, it goes back to normal.
debug.log does not say anything when it happens (or I don't know where this information could be).

**Comment by ailailou** [comment]
This is quite unfortunate. CoFH discord is throwing blame on Optifine for this issue.

For me, I'm basically stuck with pre23 atm since I'm not willing to give up either Optifine+Shaders or CoFH mods.

I'm trying out each new pre version that comes out hoping this is fixed. So far happens with pre28 as well.

I don't know how to help besides if anyone needs pre23 let me know, I was lucky to still have it in an older modpacks.

**Comment by Leirday** [comment]
The CoFH Core  Mod is also having problems with some of my other mods. (I know this isn't the place to report the issues regarding the mod.)

**Comment by TheQub** [comment]
> This is quite unfortunate. CoFH discord is throwing blame on Optifine for this issue.
> 
> For me, I'm basically stuck with pre23 atm since I'm not willing to give up either Optifine+Shaders or CoFH mods.
> 
> I'm trying out each new pre version that comes out hoping this is fixed. So far happens with pre28 as well.
> 
> I don't know how to help besides if anyone needs pre23 let me know, I was lucky to still have it in an older modpacks.

Would be awesome to try the 23,because i got no luck with newiest ones.Thanks.

**Comment by Kynolin** [comment]
@leoshusar yes that is the same problem. The screenshot in this issue is what this looks like with the internal shaders. I got black/blue/semitransparent depending on what shaders I was using.

**Comment by ailailou** [comment]
@TheQub here you go, pre23. Only issue I've noticed is the blocks breaking effects is a bit messed up but I got used to it.
[preview_OptiFine_1.15.2_HD_U_G1_pre23.jar.zip](https://github.com/sp614x/optifine/files/4808411/preview_OptiFine_1.15.2_HD_U_G1_pre23.jar.zip)

**Comment by sp614x** [comment]
The strange thing is that the bug depends on the properties of the tool used.
For OptiFine there is no difference between a stick and a wooden pickaxe, both are rendered in the same way. The tool properties (what block it can mine, etc.) don't change the rendering process.
Is CoFH trying to show some information, for example durability indicator or similar that depends on the tool properties?
Another strange thing is that the breaking animation is exactly fullscreen. It looks like the GL matrix state may be changed, for example to render some GUI info.

**Comment by KingLemming** [comment]
You're probably looking for this:

https://github.com/KingLemming/1.15/blob/master/CoFHCore/src/main/java/cofh/lib/event/AOEClientEvents.java

This file has been around for a long time, as has this functionality. It relates to drawing the block breaking graphic on all blocks in an area.

Was working with no conflicts for a long time until you updated to A26. What'd you change? ;)

**Comment by Jwguy** [comment]
This appears to still be occurring in Pre30

**Comment by PixelMellon** [comment]
I asked about this issue on the Optifine discord, referencing this article, and was just pointed back here by a bot. I can't use Optifine 1.16.1 because its incompatible with my modpack, and Pre30 still shows the bug. Would it be best to remove CoFH core dependent mods?

**Comment by QuentinSuckling** [comment]
Yeah, I've got the same issue.
I've rolled back to 1.15 Pre23 where it doesn't have that issue.
But then you miss out on all the fixes after that :/

**Comment by ailailou** [comment]
> I asked about this issue on the Optifine discord, referencing this article, and was just pointed back here by a bot. I can't use Optifine 1.16.1 because its incompatible with my modpack, and Pre30 still shows the bug. Would it be best to remove CoFH core dependent mods?

For me this was what I went for, remove CoFH mods in the end for 1.16.1, though it was a very though choice since the CoFH mods are really popular and I like them.

Hoping this will be fixed soonish.

**Comment by TheWizardLizard** [comment]
Yeah, would really liked to see this fixed.

**Comment by TheNikoWTF** [comment]
Please fix it :(

**Comment by DrJKL** [comment]
Is there any workaround in the settings?

**Comment by DrJKL** [comment]
Turning off Specular Map makes the overlay less opaque.

**Comment by Reidlos65** [comment]
interesting stalemate

**Comment by IchigoGames** [comment]
this is a issue now with 1.15.2 and recently opdated thermal expsion with opttfine 1.15.2 pre 30

**Comment by DragarX** [comment]
So are we waiting on a fix for Optifine or CoFH? because at this point downgrading optifine is not a good solution and neither team has given us news on the progress of this issue.

**Comment by IchigoGames** [comment]
most of the time optfine has to do the fix as other mod devs dont support optfine

**Comment by HarryJSJ** [comment]
Still happening on 1.16.3 - CoFH core is downloaded and I am using g4 pre2. Please fix this

**Comment by SmityJagerMan** [comment]
> Still happening on 1.16.3 - CoFH core is downloaded and I am using g4 pre2. Please fix this

Same here
OptiFine HD U G4 and G3

**Comment by KingLemming** [comment]
I've already indicated what I'm doing on my end - Optifine needs to do the same.

I have a config coming which will disable multi-block break rendering but honestly it's kind of disgraceful that I need to provide such an option.

Whatever Optifine had back in 1.15 A24 worked. It's not hard to go back and figure out what they broke. But they're closed source, so this is 100% on them to handle now.

**Comment by ailailou** [comment]
I agree with KingLemming. Bug is not gamebreaking, but would stipl like for optifine to do its part of fixing this.

**Comment by SmityJagerMan** [comment]
Well I appreciate you taking the time to even create the config option. Hopefully one day we will see an alternative to optifine. In the meantime I'll pester the optifine devs some more.

**Comment by mrgreaper** [comment]
Was a solution found to this?

**Comment by SmityJagerMan** [comment]
> Was a solution found to this?

Update cofh core and use the new config option to disable multi block breaking rendering. And tell king thanks for implementing a work around.

**Comment by mrgreaper** [comment]
> > Was a solution found to this?
> 
> Update cofh core and use the new config option to disable multi block breaking rendering. And tell king thanks for implementing a work around.

Much appreciated

**Comment by BlakeusAurelius** [comment]
Hello, I am pretty new to modern Minecraft modding as I haven't played in about 10 years. Could someone please let me know where this configuration option is to disable multi-block break rendering? I would love to be able to play Valhesia 3 with a nice shader. I updated to "cofh_core-1.16.3-1.0.4" that was released on 11/6. I am also using "OptiForge-MC1.16.3-0.3.5" & "OptiFine_1.16.3_HD_U_G4" in my mod list. Thank you in advance for any assistance and I apologize if this is a silly question!

**Comment by BlakeusAurelius** [comment]
For anyone else having trouble and that is using the twitch app here is where it is. First open your Twitch client, select Minecraft mods, then the profile you want to edit (Mine is the Valhelsia 3 option), then select the settings button in the top right -> Open Folder -> Config -> "cofh_core-client.toml". Open the "cofh_core-client.toml" with a text editor (Notepad, Notepadd ++, etc) and change the "Render Area Effect Block Breaking" value to false.

**Comment by Jfamig** [comment]
> For anyone else having trouble and that is using the twitch app here is where it is. First open your Twitch client, select Minecraft mods, then the profile you want to edit (Mine is the Valhelsia 3 option), then select the settings button in the top right -> Open Folder -> Config -> "cofh_core-client.toml". Open the "cofh_core-client.toml" with a text editor (Notepad, Notepadd ++, etc) and change the "Render Area Effect Block Breaking" value to false.

if not for this i would have never found a solution to this niche ass bug. Thank you!

**Comment by KingLemming** [comment]
It's not a solution to the bug. Optifine still hasn't addressed the actual bug. >:(

**Comment by NovaN4** [comment]
I'm also having this issue in 1.16.3. Hopefully, this bumps the thread and gets it seen if it hasn't been already, cuz I really want CoFH mods and shaders together.

**Comment by KingLemming** [comment]
Read up a couple of posts. I implemented a workaround on my end. But it's infuriating that I've had to.

**Comment by prescarlton** [comment]
@KingLemming thanks for the workaround, but sucks that you have to provide it. Side note -- also thanks for being open-source! Kudos for being a great dev.

**Comment by Cheddargt** [comment]
Is there an updated 'cofh_core-1.16.3-1.0.4.jar' with 'Render Area Effect Block Breaking = false'? @BlakeusAurelius suggested we use the Twitch App to fix this but as of 30/11/2020, the Twitch App no longer works and the CurseForge app doesn't allow for mod editing.

**Comment by loleriz** [comment]
> 
> 
> Is there an updated 'cofh_core-1.16.3-1.0.4.jar' with 'Render Area Effect Block Breaking = false'? @BlakeusAurelius suggested we use the Twitch App to fix this but as of 30/11/2020, the Twitch App no longer works and the CurseForge app doesn't allow for mod editing.
> 
> edit: lmao my bad, didn't realise this was in the game files.
> 
> For anyone wondering:
> 
> > go to .minecraft\config and edit _cofh_core-client.toml_ changing _Render Area Effect Block Breaking_ from _true_ to _false_. This can be done using text editors. Not sure about Notepad, but I used Notepad++ and it fixed it.

thank you SO MUCH for this. I was pulling my hair out trying to figure out this issue.

I tried using the preview version but that wouldn't allow structurize to load (which is needed for the valhelsia structures mod), and cofh is needed for a different mod so removing that was out of the question.

This worked perfectly, and I'm still able to use my shaders! Thanks again :)

**Comment by PenumbraStar** [comment]
This is more of a general question, as someone who has spent countless hours trying to troubleshoot optifine with mods, I am honestly curious as to why the mod remains closed source?

**Comment by KingLemming** [comment]
It sits in a weird gray area where they're breaking laws by shipping Minecraft code but Mojang is willing to let it slide apparently as long as it's discreet enough.

If Microsoft started enforcing or pursuing claims, this would be going to court immediately.

**Comment by Joaco0210** [comment]
> I've already indicated what I'm doing on my end - Optifine needs to do the same.
> 
> I have a config coming which will disable multi-block break rendering but honestly it's kind of disgraceful that I need to provide such an option.
> 
> Whatever Optifine had back in 1.15 A24 worked. It's not hard to go back and figure out what they broke. But they're closed source, so this is 100% on them to handle now.

Hey!, thanks u!, After reading this it´s very clear that the problem is optifine´s fault, but i have some information that may help to discover the cause of the bug, me and my friend have the exact same ".minecraft" folder installed that is a modification of the valhelsia 3 modpack that i made and added optifine to it, the thing is that i don´t have any problem breaking blocks, but he does, i repeat, we have the same configurations and mods so i don´t know what is the difference, if you need any extra information let me know! I will be more than happy to help.
Thanks again!!!

**Comment by KingLemming** [comment]
I've just pushed updates to CoFH Core that *should* resolve this fully, until Optifine decides to break it once again for no apparent reason.

**Comment by PenumbraStar** [comment]
Great to hear this man, you guys over at the COFH team seam to always be on top of things! Hopefully this does not occur again for a long time, thought I am not holding my breath on that one.

**Comment by sp614x** [comment]
@KingLemming what exactly is the fix?
1.15.2_pre24 adds different shaders bugfixes, some of them related to the block breaking overlay.

**Comment by KingLemming** [comment]
I had to move my code to a different event entirely and not use RenderLast (like I am supposed to do), because something in your shader fixes apparently mangles that event for the purposes of rendering block breaks.

Would be a lot easier to know what if your code was visible source.

**Comment by sp614x** [comment]
The reason is apparently the flushing of `mc.getRenderTypeBuffers().getCrumblingBufferSource()` on shaders program change or uniform change. No idea how this could lead to the breaking animation being fullscreen. 
The animation itself seems to be OK, only the matrix transform is wrong (identity?).
This is how the flush works:

        // Class IRenderTypeBuffer.Impl
        @Override
        public void flushRenderBuffers()
        {
          // Keep render type
          RenderType oldRenderType = lastRenderType;
          // Finish
          finish();
          // Restore render type
          if(oldRenderType != null)
            getBuffer(oldRenderType);
        }

Why is CoFH showing a breaking animation for the vanilla blocks? Vanilla does this already. 
Is the animation shown twice in this case?

**Comment by KingLemming** [comment]
Because I have support for tools which break blocks in an area, and it draws the breaking effect on multiple blocks at once.

**Comment by sp614x** [comment]
The problem appears with vanilla wooden pickaxe breaking vanilla stone. 
There is nothing special about them and the breaking effect is not shown on multiple blocks at once.

**Comment by diRtyMuddman** [comment]
I'm playing with the official optifine for 1.15.2, and still getting this bug on my modpack. Should I just download the updated version of cofh core for 1.15.2????

> I've just pushed updates to CoFH Core that _should_ resolve this fully, until Optifine decides to break it once again for no apparent reason.
